### PR TITLE
Us 04 quiz

### PR DESCRIPTION
**Pull request by SeacretCK**
Alpha-Version of the Quiz.
Highscores are hard-coded for now. Only the five best will be stored (in DB later). 
The nested ternary operators in the main index.js are a mess, I know. How can I do this in a better way?

### CODE REVIEW

File: src/components/US-04-Quiz/components/Highscores/index.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+              </th>
+              <th scope='col' className='px-3'>
+                Date
+              </th>
+            </tr>
+          </thead>
+          <tbody>
+            {highscores.map((entry, index) => {
+              return (
+                <tr key={index}>
```
Comment (indie-rok) [inline src/components/US-04-Quiz/components/Highscores/index.js:26]: Index should be a more generic name or you would end up with multiple elements with same index (0,1,2,3,etc). Consider adding a prefix so you have `<tr key={`table-topsongs-${index}`}

File: src/components/US-04-Quiz/components/Question/index.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        ))}
+      </div>
+      <div className='d-flex justify-content-center mt-5'>
+        <Button
+          variant='success'
+          size='lg'
+          disabled={!clicked}
+          onClick={nextQuestion}
+        >
+          {questionCounter === 9 ? 'Finish!' : 'Next'}
```
Comment (indie-rok) [inline src/components/US-04-Quiz/components/Question/index.js:84]: You can change that 9 constant into a prop like `questions.length`

File: src/components/US-04-Quiz/components/Question/index.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    const answers = [correct_answer, ...incorrect_answers];
+    const shuffledAnswers = answers.sort((a, b) => 0.5 - Math.random());
+    setAnswers(shuffledAnswers);
+  }, []);
+  // combines the correct answer with the array of wrong answers and shuffle them
+
+  const checkAnswer = (e) => {
+    if (e.target.innerText === correct_answer) {
+      e.target.className = 'btn btn-lg btn-success';
+      switch (difficulty) {
```
Comment (indie-rok) [inline src/components/US-04-Quiz/components/Question/index.js:27]: can be refactored with a hash table:

const difficultyScores = {easy:10, medium: 30, hard:50};

setScore(score + difficultyScores[difficulty])

Comment (indie-rok) [reply src/components/US-04-Quiz/components/Question/index.js:27]: This is a more declarative vs imperative way of programing btw :)

File: src/components/US-04-Quiz/components/Question/index.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        default:
+          break;
+      }
+    } else {
+      // show the clicked answer red
+      e.target.className = 'btn btn-lg btn-danger';
+      // search for the right answer and display green
+      e.target.parentNode.childNodes.forEach((answer) => {
+        if (answer.innerText === correct_answer) {
+          answer.className = 'btn btn-lg btn-success';
```
Comment (indie-rok) [inline src/components/US-04-Quiz/components/Question/index.js:46]: There is also classList.add() and classList.remove() for this purposes

File: src/components/US-04-Quiz/components/Questions/index.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+function Questions({setIsFinished, score, setScore, difficulty, setQuizStart}) {
+  const [loading, setLoading] = useState(true);
+  const [questions, setQuestions] = useState([]);
+  const [questionCounter, setQuestionCounter] = useState(0);
+  
+
+  useEffect(() => {
+    axios
+      // when difficulty is an empty string it defaults to "any", which is a mix of all three difficulty levels
+      .get(`https://opentdb.com/api.php?amount=10&category=12&difficulty=${difficulty}`)
```
Comment (indie-rok) [inline src/components/US-04-Quiz/components/Questions/index.js:15]: You can put this url into a constant

File: src/components/US-04-Quiz/components/Questions/index.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        setQuestions(results);
+        setLoading(false);
+      })
+      .catch((error) => console.log(error));
+  }, []);
+
+
+  // the API returns strings with special html entities like &#39; for apostrophes
+  // this trick inserts the string into a textarea and returns its value, thereby decoding the html characters
+  function decodeHtml(html) {
```
Comment (indie-rok) [inline src/components/US-04-Quiz/components/Questions/index.js:34]: you can put this function (and others) into a file called helpers.js or similar

File: src/components/US-04-Quiz/components/Questions/index.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    const txt = document.createElement("textarea");
+    txt.innerHTML = html;
+    return txt.value;
+}
+
+  function nextQuestion() {
+    if (questionCounter === 9) {
```
Comment (indie-rok) [inline src/components/US-04-Quiz/components/Questions/index.js:41]: let's put this 9 into a dynamic questions.length

**Review by indie-rok — COMMENTED** [review]
Great job!

I added my review only in the files with US-04-Quiz.

Let me know what you think.

File: src/components/US-04-Quiz/index.js [hunk 8]
```diff
@@ -1,5 +1,146 @@
...
+          difficulty={difficulty}
+          setQuizStart={setQuizStart}
+        />
+      ) : (
+        <>
+          {/* show startpage before and after the quiz */}
+          <StartPage start={start} setDifficulty={setDifficulty} />
+
+          {/* only after the quiz show the score */}
+          {isFinished ? (
```
Comment (EducationWildCodeSchool) [inline src/components/US-04-Quiz/index.js:93]: isFinished  && <Component >

**Review by indie-rok — APPROVED** [review]

**Pull request merged**